pkg/eks/services_v2.go

```diff
@@ -0,0 +1,55 @@
+package eks
+
+import (
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/retry"
+	"github.com/aws/aws-sdk-go-v2/service/cloudformation"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+
+	"github.com/weaveworks/eksctl/pkg/awsapi"
+)
+
+// ServicesV2 implements api.ServicesV2
+type ServicesV2 struct {
+	config aws.Config
+
+	mu             sync.Mutex
+	sts            *sts.Client
+	cloudformation *cloudformation.Client
+}
+
+// STSV2 implements the AWS STS service
+func (s *ServicesV2) STSV2() awsapi.STS {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.sts == nil {
+		s.sts = sts.NewFromConfig(s.config, func(o *sts.Options) {
+			// Disable retryer for STS
+			// (see https://github.com/weaveworks/eksctl/issues/705)
+			o.Retryer = aws.NopRetryer{}
+		})
+	}
+	return s.sts
+}
+
+// CloudFormationV2 implements the AWS CloudFormation service
+func (s *ServicesV2) CloudFormationV2() awsapi.CloudFormation {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.cloudformation == nil {
+		s.cloudformation = cloudformation.NewFromConfig(s.config, func(o *cloudformation.Options) {
+			// Use adaptive mode for retrying CloudFormation requests to mimic
+			// the logic used for AWS SDK v1.
+			o.Retryer = retry.NewAdaptiveMode(func(o *retry.AdaptiveModeOptions) {
+				o.StandardOptions = []func(*retry.StandardOptions){
+					func(so *retry.StandardOptions) {
+						so.MaxAttempts = maxRetries
+					},
+				}
+			})
+		})
+	}
+	return s.cloudformation
+}
```

Comment:
rather than doing the mutex locking and checking if we've already created the client, can we not just construct all the clients as part of `ServicesV2` and then the func's are just getters? Will reduce the complexity a bit

Reply:
> Will reduce the complexity a bit

It comes at the cost of initialising all clients even when some or none of them will be used. The approach used for SDK v1 suffers from this problem. The provider stores more than 10 clients and initialises all of them when only a few of them will be used.

Reply:
but instead we have to have this mutex and every getter be is basically a singleton constructor, this seems a lot more complicated to me

Reply:
Each client construction is fairly complex as it requires resolving multiple configurable options and setting defaults. It is wasteful to perform this initialisation for more than 10 clients when only a few of them will be used. I'll also note that the synchronisation code here can be further improved but at the cost of increased complexity.

Reply:
>  It is wasteful to perform this initialisation for more than 10 clients when only a few of them will be used. 

that doesn't concern me at all, 10 in-memory clients isn't breaking the memory bank. I'd rather have much less complex code

Reply:
My concern isn't increased memory usage, it's about wasting CPU resources initialising fields that'll never be used.

Reply:
Is that really a concern we have?

Reply:
I do, not sure about others. It's an improvement over the previous behaviour.

Reply:
> Each client construction is fairly complex as it requires resolving multiple configurable options and setting defaults. It is wasteful to perform this initialisation for more than 10 clients when only a few of them will be used. I'll also note that the synchronisation code here can be further improved but at the cost of increased

+1 on this, i like this approach.

Reply:
I really disagree, the code is much more complex this way, all to save the costs of creating a handful more structs? :confused:

Reply:
Lets take this offline 👍🏻

Reply:
I can see both sides :smile: 

In reality, we don't need to worry too much about optimizing this area and creating all the services would be fine.

On the flip side, its a standard lazy initialization pattern. The one thing i would say is that we are assuming that all new services will use this pattern and that contributors will need to follow it. So, it might be worth documenting it somewhere?

Reply:
Do we have any kind of benchmarking that supports this "optimisation"? Because without it, this is just going in blind. At first, it actually causes a problem then solve one which might not exist.

Further, this will remove the nice UX of failing early in case the client can't be constructed instead of much later in the process.

And, I think this should have been in a separate PR and not added to a Pr which already huge.

I don't agree with these changes. :/

Reply:
Oh see, this is just constructing structs? I would argue that this would have impacted performance in any way. I would be very surprised. Which is why a benchmark should have existed first. ¯\_(ツ)_/¯

Reply:
>Do we have any kind of benchmarking that supports this "optimisation"? Because without it, this is just going in blind.

Please go through the entire discussion on this topic. I have already commented this: "this change was not just about optimizing code in terms of CPU cycles, but also to communicate the intent that only a few of these services will be used by most code paths."

This changelist isn't optimizing existing code, it's making sure that no code is unnecessarily run if it's not used.


>Further, this will remove the nice UX of failing early in case the client can't be constructed instead of much later in the process.

Where exactly was it failing early before? The `NewFromConfig` functions do not return an error.

>Oh see, this is just constructing structs? I would argue that this wouldn't have impacted performance in any way. I would be very surprised. Which is why a benchmark should have existed first. ¯\_(ツ)_/¯

"Constructing structs" happens at the end when all the initialization and resolution functions have been run. To "just construct structs" you need to set its fields which requires making several function calls here in the `NewFromConfig` functions. But as I noted above, this is not just trying to improve its performance.

Reply:
Which would have been more evident if this entire change set would have been a separate PR where all the moving and related parts were visible. I'm not against doing lazy loading, but I was against doing it in an already huge pr with several changes in it, related or not. This should have been in a separate pr for clear visibility.